programs/host/frontend.c

```diff
@@ -0,0 +1,957 @@
+            {
+                mbedtls_net_context ctx = { item_uint16( inputs[0] ) };
+                uint32_t len = item_uint32( inputs[1] );
+                uint32_t timeout = item_uint32( inputs[2] );
+                ALLOC_OUTPUT( 0 , len ); // data
+                if( timeout == MBEDTLS_SERIALIZE_TIMEOUT_INFINITE )
+                {
+                    DBG( "executing recv fd=%u len=%u", (int) ctx.fd, (unsigned) len );
+                    ret = mbedtls_net_recv( &ctx, item_buffer( outputs[0] ), len );
+                }
+                else
+                {
+                    DBG( "executing recv_timeout fd=%u len=%u timeout=%u",
+                        (unsigned) ctx.fd, (unsigned) len, (unsigned) timeout );
+                    ret = mbedtls_net_recv_timeout( &ctx,
+                            item_buffer( outputs[0] ), len,
+                            timeout );
+                }
+                if( ret >= 0 )
+                {
+                    DBG( "received %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                    outputs[0]->size = ret;
+                    ret = 0;
+                }
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SEND:
+            CHECK_ARITY( 2 );
+            CHECK_LENGTH( 0, 2 ); // fd
+            {
+                mbedtls_net_context ctx = { item_uint16( inputs[0] ) };
+                size_t len = inputs[1]->size;
+                unsigned char *buf = item_buffer( inputs[1] );
+                ALLOC_OUTPUT( 0 , 4 ); // sent_len
+                DBG( "executing send fd=%u len=%zu", (int) ctx.fd, len );
+                ret = mbedtls_net_send( &ctx, buf, len );
+                if( ret >= 0 )
+                {
+                    DBG( "sent %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                    set_item_uint32( outputs[0], ret );
+                    ret = 0;
+                }
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SHUTDOWN:
+            CHECK_ARITY( 1 );
+            CHECK_LENGTH( 0, 2 ); // fd
+            {
+                mbedtls_net_context ctx = { item_uint16( inputs[0] ) };
+                DBG( "executing shutdown fd=%d", (int) ctx.fd );
+                mbedtls_net_free( &ctx );
+                ret = 0;
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_FOPEN:
+            {
+                char * path = NULL;
+                char * mode = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                mode = item_buffer( inputs[0] );
+                path = item_buffer( inputs[1] );
+                DBG( "open file [%s] mode [%s]", path, mode);
+                file_id = alloc_file_context();
+                DBG( "allocated file id [%d]", file_id);
+                if ( file_id != -1 )
+                {
+                    FILE * file = fopen( path, mode );
```

Comment:
We should use `mbedtls_fopen` rather than `fopen` here. The frontend is meant to run on “full” platforms, so we can assume a hosted C implementation, i.e. we can assume that all standard library functions are present. If someone has overridden the implementation of `mbedtls_fopen` for the platform they're building the frontend on, then they must have a good reason and we shouldn't bypass it.

Reply:
mmm, frontend.c is not a consumer of the serialisation. Rather it serves the serialisation by called platform specific ```fopen```. Function ```mbedtls_fopen``` when called from the library (with serialisation turned on) results in above piece of code being executed. If I use the same function here wouldn't it cause a recursion?

Reply:
@mazimkhan `frontend.c` is compiled with a different `config.h`. It has to, since settings like `MBEDTLS_PLATFORM_xxx` and `MBEDTLS_HAVE_xxx` need to be for the host platform, not for the target platform. This is how the network functions work.

Reply:
Yes I know that. However, it doesn't/shouldn't require the library apart from ```serialize.h``` that is the interface specification. Specially for ```mbedtls_fopen``` and ```mbedtls_opendir``` that are simply hash defined to ```fopen``` and ```opendir```.  I know flags makes it possible. But I find it uneasy.
It adds circular dependency library->frontend->library. frontend could be completely independent and does not have to tie to the library apart from the specification.